File: lightroom_tagger/core/identity_service/ranking.py
"""Best-photo ranking with stack enrichment."""

from __future__ import annotations

import sqlite3
from typing import Any

from .percentiles import compute_image_peak_percentile_scores


def _image_meta_map(conn: sqlite3.Connection, keys: list[str]) -> dict[str, dict[str, Any]]:
    if not keys:
        return {}
    placeholders = ",".join("?" * len(keys))
    out: dict[str, dict[str, Any]] = {}
    # Catalog images
    rows = conn.execute(
        f"SELECT key, filename, date_taken, rating, instagram_posted FROM images "
        f"WHERE key IN ({placeholders})",
        keys,
    ).fetchall()
    for r in rows:
        out[str(r["key"])] = {
            "filename": r.get("filename") or "",
            "date_taken": r.get("date_taken") or "",
            "rating": int(r["rating"] or 0),
            "instagram_posted": bool(r.get("instagram_posted")),
            "image_type": "catalog",
        }
    # Instagram images (keys not already resolved above)
    missing = [k for k in keys if k not in out]
    if missing:
        ig_placeholders = ",".join("?" * len(missing))
        ig_rows = conn.execute(
            f"SELECT media_key, filename, created_at FROM instagram_dump_media "
            f"WHERE media_key IN ({ig_placeholders})",
            missing,
        ).fetchall()
        for r in ig_rows:
            out[str(r["media_key"])] = {
                "filename": r.get("filename") or r.get("media_key") or "",
                "date_taken": r.get("created_at") or "",
                "rating": 0,
                "instagram_posted": True,
                "image_type": "instagram",
            }
    return out


def _stack_non_representative_keys(conn: sqlite3.Connection, keys: list[str]) -> set[str]:
    """Image keys that are stack members but not the stack representative."""
    if not keys:
        return set()
    placeholders = ",".join("?" * len(keys))
    rows = conn.execute(
        f"""
        SELECT m.image_key FROM image_stack_members m
        INNER JOIN image_stacks s ON s.stack_id = m.stack_id
        WHERE m.image_key IN ({placeholders}) AND m.image_key <> s.representative_key
        """,
        keys,
    ).fetchall()
    return {str(r["image_key"]) for r in rows}


def _stack_fields_for_image_keys(
    conn: sqlite3.Connection, keys: list[str]
) -> dict[str, dict[str, Any]]:
    """``stack_id``, ``stack_member_count`` (``image_stacks.stack_size``), ``is_stack_representative``."""
    if not keys:
        return {}
    placeholders = ",".join("?" * len(keys))
    rows = conn.execute(
        f"""
        SELECT m.image_key, s.stack_id, s.representative_key, s.stack_size
        FROM image_stack_members m
        INNER JOIN image_stacks s ON s.stack_id = m.stack_id
        WHERE m.image_key IN ({placeholders})
        """,
        keys,
    ).fetchall()
    out: dict[str, dict[str, Any]] = {}
    for r in rows:
        k = str(r["image_key"])
        rep = str(r["representative_key"])
        out[k] = {
            "stack_id": int(r["stack_id"]),
            "stack_member_count": int(r["stack_size"]),
            "is_stack_representative": k == rep,
        }
    return out


def rank_best_photos(
    conn: sqlite3.Connection,
    *,
    limit: int,
    offset: int,
    min_perspectives: int | None = None,
    sort_by_date: str | None = None,
    posted: bool | None = None,
) -> tuple[list[dict[str, Any]], int, dict[str, Any]]:
    """Eligible images only, sorted by peak_percentile DESC, date_taken, key ASC.

    ``sort_by_date`` (``newest`` / ``oldest``) only controls the date tiebreaker;
    peak within-perspective percentile remains the primary sort key.
    """
    if sort_by_date is not None and sort_by_date not in ("newest", "oldest"):
        raise ValueError("sort_by_date must be 'newest' or 'oldest'")

    items, meta = compute_image_peak_percentile_scores(
        conn, min_perspectives=min_perspectives, include_ineligible=False
    )
    eligible = [i for i in items if i.get("eligible")]
    keys = [str(i["image_key"]) for i in eligible]
    img_meta = _image_meta_map(conn, keys)

    enriched: list[dict[str, Any]] = []
    for i in eligible:
        k = str(i["image_key"])
        im = img_meta.get(k, {})
        enriched.append({**i, **im})

    ekeys = [str(r["image_key"]) for r in enriched]
    drop_keys = _stack_non_representative_keys(conn, ekeys)
    enriched = [r for r in enriched if str(r["image_key"]) not in drop_keys]

    if enriched:
        skeys = [str(r["image_key"]) for r in enriched]
        stack_by_key = _stack_fields_for_image_keys(conn, skeys)
        for r in enriched:
            k = str(r["image_key"])
            if k in stack_by_key:
                r.update(stack_by_key[k])
            else:
                r["stack_id"] = None
                r["stack_member_count"] = None
                r["is_stack_representative"] = False

    date_reverse = sort_by_date != "oldest"
    enriched.sort(key=lambda r: r["image_key"])
    enriched.sort(key=lambda r: r.get("date_taken") or "", reverse=date_reverse)
    enriched.sort(key=lambda r: r["peak_percentile"], reverse=True)

    if posted is True:
        enriched = [r for r in enriched if bool(r.get("instagram_posted")) is True]
    elif posted is False:
        enriched = [r for r in enriched if bool(r.get("instagram_posted")) is False]

    total = len(enriched)
    page = enriched[offset : offset + limit]
    return page, total, meta

File: apps/visualizer/backend-ts/src/identity/ranking.ts
/**
 * Best-photo ranking with stack enrichment.
 * Port of `core/identity_service/ranking.py`.
 */
import type { Db } from '../db/connection.js';
import { computeImagePeakPercentileScores, type PeakPercentileItem, type PeakPercentileMeta } from './percentiles.js';

/**
 * Maximum keys per `IN (...)` list.
 *
 * These helpers are called with *every* scored image in the catalog — the Mirror
 * passes all ~29,000 at once — and one bound parameter per key blows SQLite's
 * variable limit. Python got away with it because CPython's bundled SQLite is built
 * with a high `SQLITE_MAX_VARIABLE_NUMBER`; better-sqlite3's answers "too many SQL
 * variables". 900 is under the 999 floor that any SQLite build guarantees.
 */
const KEY_CHUNK = 900;

/** Run `query` over `keys` in batches, concatenating the rows. */
function inChunks<T>(keys: readonly string[], query: (batch: string[]) => T[]): T[] {
  const out: T[] = [];
  for (let i = 0; i < keys.length; i += KEY_CHUNK) {
    out.push(...query([...keys.slice(i, i + KEY_CHUNK)]));
  }
  return out;
}

export interface ImageMeta {
  filename: string;
  date_taken: string;
  rating: number;
  instagram_posted: boolean;
  image_type: 'catalog';
}

export interface StackFields {
  stack_id: number | null;
  stack_member_count: number | null;
  is_stack_representative: boolean;
}

/** Catalog metadata for the ranked keys, keyed by image key. */
export function imageMetaMap(db: Db, keys: readonly string[]): Map<string, ImageMeta> {
  if (keys.length === 0) return new Map();
  const out = new Map<string, ImageMeta>();
  const rows = inChunks(keys, (batch) =>
    db
      .prepare(
        `SELECT key, filename, date_taken, rating, instagram_posted FROM images ` +
          `WHERE key IN (${batch.map(() => '?').join(',')})`,
      )
      .all(...batch) as {
      key: string;
      filename: string | null;
      date_taken: string | null;
      rating: number | null;
      instagram_posted: number | null;
    }[],
  );
  for (const r of rows) {
    out.set(String(r.key), {
      filename: r.filename || '',
      date_taken: r.date_taken || '',
      rating: Math.trunc(r.rating || 0),
      instagram_posted: Boolean(r.instagram_posted),
      image_type: 'catalog',
    });
  }
  return out;
}

/**
 * Keys that are stack members but not the representative.
 *
 * These are dropped from every identity ranking: a burst of twenty near-identical
 * frames would otherwise fill the whole page.
 */
export function stackNonRepresentativeKeys(db: Db, keys: readonly string[]): Set<string> {
  if (keys.length === 0) return new Set();
  const rows = inChunks(keys, (batch) =>
    db
      .prepare(
        `
        SELECT m.image_key FROM image_stack_members m
        INNER JOIN image_stacks s ON s.stack_id = m.stack_id
        WHERE m.image_key IN (${batch.map(() => '?').join(',')}) AND m.image_key <> s.representative_key
        `,
      )
      .all(...batch) as { image_key: string }[],
  );
  return new Set(rows.map((r) => String(r.image_key)));
}

/**
 * Stack columns for the given keys.
 *
 * `stack_member_count` comes from the denormalized `image_stacks.stack_size` here,
 * unlike `stackMetadataForApi`, which counts live membership. Kept as-is so the
 * identity payload matches what Flask sent.
 */
export function stackFieldsForImageKeys(
  db: Db,
  keys: readonly string[],
): Map<string, StackFields> {
  if (keys.length === 0) return new Map();
  const rows = inChunks(keys, (batch) =>
    db
      .prepare(
        `
        SELECT m.image_key, s.stack_id, s.representative_key, s.stack_size
        FROM image_stack_members m
        INNER JOIN image_stacks s ON s.stack_id = m.stack_id
        WHERE m.image_key IN (${batch.map(() => '?').join(',')})
        `,
      )
      .all(...batch) as {
      image_key: string;
      stack_id: number;
      representative_key: string;
      stack_size: number;
    }[],
  );
  const out = new Map<string, StackFields>();
  for (const r of rows) {
    const k = String(r.image_key);
    out.set(k, {
      stack_id: Math.trunc(r.stack_id),
      stack_member_count: Math.trunc(r.stack_size),
      is_stack_representative: k === String(r.representative_key),
    });
  }
  return out;
}

export type RankedItem = PeakPercentileItem & Partial<ImageMeta> & Partial<StackFields>;

/** Compare two strings the way Python's `<` does for ASCII keys. */
function cmpStr(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

/**
 * Eligible images only, ordered by `ranking_percentile` descending.
 *
 * `sortByDate` only moves the *tiebreaker*: the corroboration-vetoed ranking
 * percentile stays the primary key, so asking for "oldest" reorders equally-ranked
 * photos rather than turning this into a date listing.
 *
 * The three successive stable sorts reproduce Python's exactly — key ascending,
 * then date, then ranking — which is not the same as one comparator with three
 * clauses when equal ranking values arise from different rounding paths.
 */
export function rankBestPhotos(
  db: Db,
  opts: {
    limit: number;
    offset: number;
    minPerspectives?: number | null;
    sortByDate?: 'newest' | 'oldest' | null;
    posted?: boolean | null;
  },
): { items: RankedItem[]; total: number; meta: PeakPercentileMeta } {
  const sortByDate = opts.sortByDate ?? null;
  if (sortByDate !== null && sortByDate !== 'newest' && sortByDate !== 'oldest') {
    throw new RangeError("sort_by_date must be 'newest' or 'oldest'");
  }

  const { items, meta } = computeImagePeakPercentileScores(db, {
    minPerspectives: opts.minPerspectives ?? null,
    includeIneligible: false,
  });
  const eligible = items.filter((i) => i.eligible);
  const imgMeta = imageMetaMap(db, eligible.map((i) => String(i.image_key)));

  let enriched: RankedItem[] = eligible.map((i) => ({
    ...i,
    ...(imgMeta.get(String(i.image_key)) ?? {}),
  }));

  const dropKeys = stackNonRepresentativeKeys(db, enriched.map((r) => String(r.image_key)));
  enriched = enriched.filter((r) => !dropKeys.has(String(r.image_key)));

  if (enriched.length > 0) {
    const stackByKey = stackFieldsForImageKeys(db, enriched.map((r) => String(r.image_key)));
    for (const r of enriched) {
      const fields = stackByKey.get(String(r.image_key));
      if (fields) Object.assign(r, fields);
      else {
        r.stack_id = null;
        r.stack_member_count = null;
        r.is_stack_representative = false;
      }
    }
  }

  const dateDescending = sortByDate !== 'oldest';
  enriched.sort((a, b) => cmpStr(String(a.image_key), String(b.image_key)));
  enriched.sort((a, b) => {
    const c = cmpStr(a.date_taken ?? '', b.date_taken ?? '');
    return dateDescending ? -c : c;
  });
  enriched.sort((a, b) => b.ranking_percentile - a.ranking_percentile);

  if (opts.posted === true) enriched = enriched.filter((r) => r.instagram_posted === true);
  else if (opts.posted === false) enriched = enriched.filter((r) => !r.instagram_posted);

  const total = enriched.length;
  return { items: enriched.slice(opts.offset, opts.offset + opts.limit), total, meta };
}
